lib/std/crypto/timing_safe.zig

```diff
@@ -131,6 +131,56 @@ pub fn sub(comptime T: type, a: []const T, b: []const T, result: []T, endian: En
     return @as(bool, @bitCast(borrow));
 }
 
+fn markSecret(ptr: anytype, comptime action: enum { classify, declassify }) void {
+    const t = @typeInfo(@TypeOf(ptr));
+    if (t != .Pointer) @compileError("Pointer expected - Found: " ++ @typeName(@TypeOf(ptr)));
+    const p = t.Pointer;
+    if (p.is_allowzero) @compileError("A nullable pointer is always assumed to leak information via side channels");
+    const child = @typeInfo(p.child);
+
+    switch (child) {
+        .Void, .Null, .ComptimeInt, .ComptimeFloat => return,
+        .Pointer => {
+            if (child.Pointer.size == std.builtin.Type.Pointer.Size.Slice) {
+                @compileError("Found pointer to pointer. If the intent was to pass a slice, maybe remove the leading & in the function call");
+            }
+            @compileError("A pointer value is always assumed leak information via side channels");
+        },
+        else => {
+            const mem8: *const [@sizeOf(@TypeOf(ptr.*))]u8 = @constCast(@ptrCast(ptr));
+            if (action == .classify) {
+                std.valgrind.memcheck.makeMemUndefined(mem8);
+            } else {
+                std.valgrind.memcheck.makeMemDefined(mem8);
+            }
+        },
+    }
+}
+
+/// Mark a value as sensitive or secret, helping to detect potential side-channel vulnerabilities.
+///
+/// When Valgrind is enabled, this function allows for the detection of conditional jumps or lookups
+/// that depend on secrets or secret-derived data. Violations are reported by Valgrind as operations
+/// relying on uninitialized values.
+///
+/// If Valgrind is disabled, it has no effect.
+///
+/// Use this function to verify that cryptographic operations perform constant-time arithmetic on sensitive data,
+/// ensuring the confidentiality of secrets and preventing information leakage through side channels.
+pub fn classify(ptr: anytype) void {
+    if (!std.debug.inValgrind()) return;
+    markSecret(ptr, .classify);
+}
+
+/// Mark a value as non-sensitive or public, indicating it's safe from side-channel attacks.
+///
+/// Signals that a value has been securely processed and is no longer confidential, allowing for
+/// relaxed handling without fear of information leakage through conditional jumps or lookups.
+pub fn declassify(ptr: anytype) void {
+    if (!std.debug.inValgrind()) return;
```

Comment:
This does both a runtime and a compile-time check; where the runtime check is as expensive as just doing the operation anyway.
i.e. no need for this check: the `makeMemUndefined`/`makeMemDefined` compile to no-ops if not running under valgrind
Though perhaps you should add in a check for debug/safe vs release-fast?

Reply:
> i.e. no need for this check: the makeMemUndefined/makeMemDefined compile to no-ops

Is that true without LTO?

Reply:
The `std.debug.inValgrind()` checks have been removed.

A check for debug/safe vs release-fast can be useful if `classify()` does a bit more than call valgrind, but this is not the case yet.